jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java

```diff
@@ -581,31 +581,34 @@ public void parse(final Set<? extends Handler> handlers, Resource r) throws Exce
      */
     protected void parseDir(Set<? extends Handler> handlers, Resource dirResource) throws Exception
     {
-        Path dir = dirResource.getPath();
-
         if (LOG.isDebugEnabled())
-            LOG.debug("Scanning dir {}", dir.toUri());
+            LOG.debug("Scanning dir {}", dirResource);
 
         ExceptionUtil.MultiException multiException = new ExceptionUtil.MultiException();
-        try (Stream<Path> classStream = Files.walk(dir))
-        {
-            classStream
-                .filter(Files::isRegularFile)
-                .filter((path) -> !FileID.isHidden(dir, path))
-                .filter(FileID::isNotMetaInfVersions)
-                .filter(FileID::isNotModuleInfoClass)
-                .filter(FileID::isClassFile)
-                .forEach(classFile ->
-                {
-                    try
-                    {
-                        parseClass(handlers, dirResource, classFile);
-                    }
-                    catch (Exception ex)
+        for (Resource r : dirResource)
+        {
+            Path dir = r.getPath();
+
+            try (Stream<Path> classStream = Files.walk(dir))
+            {
```

Comment:
This fixes the issue, but not correctly.

By iterating over any combined resources, the parser might see multiple versions of the same class from different layers in an overlay, when it should only see the topmost one.
The solution is that we should not be using `Path.walk`.   Instead we need to implement a `Resource.walk` method that correctly handles duplicates in an overlay.